### PR TITLE
fix: Cleanup sticky columns listeners when table cells unmount

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

The missing subscription cleanup function caused a memory leak.

ref: AWSUI-22191

### How has this been tested?

New unit test added.

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/table/sticky-columns/__tests__/use-sticky-columns.test.tsx [hunk 1]
```diff
@@ -241,6 +241,38 @@ test('performs styles cleanup', () => {
   expect(elements.cells[0]).not.toHaveClass('sticky-cell');
 });
 
+test('cell subscriptions are cleaned up on ref change and un-mount', () => {
+  const unsubscribe = jest.fn();
+  const subscribe = jest.fn().mockImplementation(() => unsubscribe);
```
Comment (connorlanigan) [inline src/table/sticky-columns/__tests__/use-sticky-columns.test.tsx:246]: just fyi, you can pass the mock implementation directly as a parameter:
```js
jest.fn(() => unsubscribe)
```

File: src/table/sticky-columns/use-sticky-columns.ts [hunk 2]
```diff
@@ -201,6 +201,11 @@ export function useStickyCellStyles({
     [columnId, setCell, stickyColumns.store]
   );
 
+  // Subscriptions created in the refCallback must be cancelled when the component un-mounts.
+  useEffect(() => {
+    return () => unsubscribeRef.current?.();
```
Comment (connorlanigan) [inline src/table/sticky-columns/use-sticky-columns.ts:206]: Looks good, but I don't quite understand yet why we need to do this. Shouldn't react call the refCallback with `null` when the component unmounts, and thus already cause the subscription to be cancelled in line 162?

Comment (pan-kot) [reply src/table/sticky-columns/use-sticky-columns.ts:206]: Great question! That should indeed be the case.

Comment (pan-kot) [reply src/table/sticky-columns/use-sticky-columns.ts:206]: Updated!

File: src/table/body-cell/td-element.tsx [hunk 3]
```diff
@@ -97,11 +97,9 @@ export const TableTdElement = React.forwardRef<HTMLTableCellElement, TableTdElem
         onMouseEnter={onMouseEnter}
         onMouseLeave={onMouseLeave}
         ref={node => {
-          if (node) {
-            stickyStyles.ref(node);
-            if (ref) {
-              (ref as React.MutableRefObject<HTMLTableCellElement>).current = node;
-            }
+          stickyStyles.ref(node);
+          if (node && ref) {
```
Comment (connorlanigan) [inline src/table/body-cell/td-element.tsx:101]: Can we remove the `ref` condition for this other ref as well? This way we will match the usual React expectations of refs being set to `null` when a component is unmounted
i.e.:
```
ref={node => {
  stickyStyles.ref(node);
  if (ref) {
    (ref as React.MutableRefObject<HTMLTableCellElement>).current = node;
  }
}

File: src/table/sticky-columns/__tests__/use-sticky-columns.test.tsx [hunk 4]
```diff
@@ -267,7 +265,7 @@ test('cell subscriptions are cleaned up on ref change and un-mount', () => {
   expect(subscribe).toHaveBeenCalledTimes(2);
   expect(unsubscribe).toHaveBeenCalledTimes(1);
 
-  unmount();
+  result.current.ref(null);
```
Comment (connorlanigan) [inline src/table/sticky-columns/__tests__/use-sticky-columns.test.tsx:268]: Does this test still work if we use `unmount`? Internally the `unmount` call would then still call `result.current.ref(null)`, but it expresses a bit clearer what we want to simulate in this line than just calling our ref function directly

**Review by connorlanigan — CHANGES_REQUESTED** [review]

**Review by connorlanigan — APPROVED** [review]

**Pull request merged**